_posts/2018-11-16-grpc-dotnet-build.md

```diff
@@ -0,0 +1,175 @@
+
+kkm@yupana:~/work$ cd MyGreeter
+kkm@yupana:~/work/MyGreeter$ ls -lF
+total 12
+-rw-rw-r-- 1 kkm kkm   86 Nov  9 16:10 Class1.cs
+-rw-rw-r-- 1 kkm kkm  145 Nov  9 16:10 MyGreeter.csproj
+drwxrwxr-x 2 kkm kkm 4096 Nov  9 16:10 obj/
+```
+
+The `dotnet new` command has created the file `Class1.cs` that we won't need, so
+it's ok to remove it. Also, we need some `.proto` files to compile. For this
+exercise, we'll copy an example file [`examples/protos/helloworld.proto`
+](https://github.com/grpc/grpc/blob/master/examples/protos/helloworld.proto)
+from the gRPC distribution.
+
+```
+kkm@yupana:~/work/MyGreeter$ rm Class1.cs
+kkm@yupana:~/work/MyGreeter$ wget -q https://raw.githubusercontent.com/grpc/grpc/master/examples/protos/helloworld.proto
+```
+
+(on Windows, use `del Class1.cs`, and, if you do not have the wget command,
+just [open the above URL
+](https://raw.githubusercontent.com/grpc/grpc/master/examples/protos/helloworld.proto)
+and use a *Save As...* command from your Web browser).
+
+Next, add required NuGet packages to the project:
+
+```
+kkm@yupana:~/work/MyGreeter$ dotnet add package Grpc
+info : PackageReference for package 'Grpc' version '1.17.0' added to file '/home/kkm/work/MyGreeter/MyGreeter.csproj'.
+kkm@yupana:~/work/MyGreeter$ dotnet add package Grpc.Tools
+info : PackageReference for package 'Grpc.Tools' version '1.17.0' added to file '/home/kkm/work/MyGreeter/MyGreeter.csproj'.
+kkm@yupana:~/work/MyGreeter$ dotnet add package Google.Protobuf
+info : PackageReference for package 'Google.Protobuf' version '3.6.1' added to file '/home/kkm/work/MyGreeter/MyGreeter.csproj'.
+```
+
+**Next comes an important part.** First of all, by default, a `.csproj` project
+file automatically finds all `.cs` files in its directory, although
+[Microsoft now recommends suppressing this globbing
+behavior](https://docs.microsoft.com/dotnet/core/tools/csproj#recommendation),
+so we too decided against globbing `.proto` files. Thus the `.proto`
+files must be added to the project explicitly. Second of all, it is important
+to add a property `PrivateAssets="All"` to the Grpc.Tools package reference,
+so that it will not be needlessly fetched by the consumers of your new library.
+This makes sense, as the package only contains compiler, code generators and
+import files, which are not needed outside of the project where the `.proto`
+files have been compiled. So you must edit the file `MyGreeter.csproj` and
+add the `helloworld.proto` to be compiled, and the `PrivateAssets` property to
+the Grpc.Tools package reference. Your resulting project file should now look
+like this:
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Google.Protobuf" Version="3.6.1" />
+    <PackageReference Include="Grpc" Version="1.17.0" />
+    <PackageReference Include="Grpc.Tools" Version="1.17.0" PrivateAssets="All" />
```

Comment:
Add XML comments highlighting what are the two most important parts in this file?   (=adding  `PrivateAssets="All"`  and adding `<ProtoBuf Include="helloworld.proto" />`. The rest of the file is basically boilerplate.